### DISCUSSION TITLE
consider switching to byte(s) notation for primitive types, instead of bits

### DISCUSSION THREAD

**Discussion by gintominto5329** [carbon-language/carbon-lang#3602] [Language design]
example : `u8`(unsigned integer ,8 bytes) vs. `u64`(unsigned integer ,64 bits) ,`u1` vs. `u8`

Because bits(un-aligned to a byte ,or power of 2) are highly unlikely to come back.

and the bits usage is merely historical

**Comment by jonmeow** [comment]
The specific `i32` naming scheme was chosen to be consistent Rust; C++ uses `int32_t` and similar, while Swift uses `Int32`. Bits show up in many places in computing: CPUs (x86-64, amd64), OS (64-bit Windows), checksums (sha256), etc. Disk and RAM are typically in bytes, although bandwidth again in bits. Familiarity is important; I'm not aware of any programming language that defines something such as `Int4` as 4-bytes.

There's also a question of what it means for transitioning developers. In a byte-based syntax, `i8` (64-bit), `i16` (128-bit), `i32` (256-bit), and `i64` (512-bit) would all be valid, [per design](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/expressions/literals.md#numeric-type-literals) to better support large integers. As a consequence, if a developer writes code mis-applying the familiar bit syntax, the translation between bits and bytes would likely result in valid but bug-prone programs. For example, `var a: i32 = 1 << 30;` is valid regardless of whether `i32` is 32 bits or 32 bytes, but performance bugs should be expected if that's 32 bytes. Runtime crashes could result too, for example if someone incorrectly assumes `i8` is only 8 bits and can safely be cast to a `Char`.

Elsewhere within the syntax, bit shift operators will also remain prevalent.

Leaning into common precedent takes advantage of existing expectations in order to create an easier learning curve. Both bit and byte syntax are common in computing, but there's a lot of value in being able to say: on a 64-bit CPU with a 64-bit OS, where someone might use `int64_t` in C++, use `i64` in Carbon.

**Reply by gintominto5329** [reply]
I understand your point ,of familiarity ,and psychology.

But ,may i have your opinion ,on ,is byte-based naming more correct ,now?

thanks

**Reply by chandlerc** [reply]
I don't think it's any more (or less) correct.

As was mentioned, there remain lots of uses of bits to directly articulate width.
